lib/index.js

```diff
@@ -156,38 +156,50 @@ internals.implementation = function (server, options) {
 
                 if (!settings.validateFunc) {
                     if (settings.keepAlive) {
-                        reply.state(settings.cookie, session);
+                        h.state(settings.cookie, session);
                     }
 
-                    return reply.continue({ credentials: session, artifacts: session });
+                    return h.authenticated({ credentials: session, artifacts: session });
                 }
 
-                settings.validateFunc(request, session, (err, isValid, credentials) => {
+                let credentials = session;
 
-                    if (err ||
-                        !isValid) {
+                try {
+                    const result = await settings.validateFunc(request, session);
 
-                        if (settings.clearInvalid) {
-                            reply.unstate(settings.cookie);
-                        }
+                    Hoek.assert(typeof result === 'object', 'Invalid return from validateFunc');
```

Comment:
Here you're validating the user's own code. It feels a bit out of scope, they should have tests.

Reply:
Personally I find it useful that the framework verifies the results of user provided hooks for ease of debugging, unless the benefits are outweighed by performance concerns.

Reply:
There's no harm in it I guess. I don't think there's any perf concerns with a couple of asserts.